colcon_core/extension_point.py

```diff
@@ -44,27 +50,25 @@ def get_all_extension_points():
     colcon_extension_points.setdefault(EXTENSION_POINT_GROUP_NAME, None)
 
     entry_points = defaultdict(dict)
-    working_set = WorkingSet()
-    for dist in sorted(working_set):
-        entry_map = dist.get_entry_map()
-        for group_name in entry_map.keys():
+    seen = set()
+    for dist in distributions():
```

Comment:
As far as I understand, this no longer works as expected in the current code due to pypa/setuptools#3791, and the importlib-based class isn't inherently sort-able. We'd have to implement our own sort routine, which would involve parsing the version numbers for comparison. It has the potential to be pretty expensive.

From what I've gathered, it's not a well-supported scenario by any tooling to have multiple versions of a distribution installed in the same prefix anyway.

Also worth questioning is whether this was the right thing to do to begin with. Assuming the working set contained all of the distributions across all of the discoverable prefixes, sorting it like that would always prefer the newest version of a distribution, where an import would prefer the one(s) discoverable soonest in the prefix list.

If I'm not mistaken, the existing code wouldn't allow you to (for example) build an older version of a package in a colcon workspace than the one installed on your system and source that workspace, because this code would always prefer the newer version even though an `import` statement would not.

Reply:
I dug into this a little more.

In `pkg_resources`, it appears that the path ordering was respected and you were never given multiple distributions with the same "key", so if there are multiple dists in the same path directory, it's a matter of luck which one you get.

In `importlib.metadata`, the dists are not de-duplicated, so you will likely encounter multiple with the same key. If we sorted them, you might get "inferior" or shadowed versions showing up with higher precedence because they have a higher version number.

Since multiple versions in the same path really isn't supported well anywhere, I don't think there's any harm in skipping the sort for performance reasons. I'll say that the one remaining benefit is that it makes the duplicate entry point definition error more deterministic.